internal/controlplane/handlers_oauth.go

```diff
@@ -115,55 +127,93 @@ func (s *Server) GetAuthorizationURL(ctx context.Context,
+// note: this is an HTTP only (not RPC) handler
+func (s *Server) HandleProviderCallback() runtime.HandlerFunc {
+	return func(w http.ResponseWriter, r *http.Request, pathParams map[string]string) {
+		ctx := context.Background()
+
+		if err := s.processCallback(ctx, w, r, pathParams); err != nil {
+			log.Printf("error handling provider callback: %s", err)
+			w.WriteHeader(http.StatusInternalServerError)
+			return
+		}
+	}
+}
+
+func (s *Server) processCallback(ctx context.Context, w http.ResponseWriter, r *http.Request,
+	pathParams map[string]string) error {
+	state := r.URL.Query().Get("state")
+	code := r.URL.Query().Get("code")
 
 	// Configure tracing
 	span := trace.SpanFromContext(ctx)
-	span.SetName("server.ExchangeCodeForTokenCLI")
-	span.SetAttributes(attribute.Key("code").String(in.Code))
+	span.SetName("server.HandleProviderCallback")
+	span.SetAttributes(attribute.Key("code").String(code))
 	defer span.End()
 
-	// Check the nonce to make sure it's valid
-	valid, err := mcrypto.IsNonceValid(in.State, s.cfg.Auth.NoncePeriod)
+	provider := pathParams["provider"]
 
+	// Check the nonce to make sure it's valid
+	valid, err := mcrypto.IsNonceValid(state, s.cfg.Auth.NoncePeriod)
 	if err != nil {
-		return nil, status.Errorf(codes.Unknown, "error checking nonce: %s", err)
+		return fmt.Errorf("error checking nonce: %w", err)
 	}
-
 	if !valid {
-		return nil, status.Error(codes.InvalidArgument, "invalid nonce")
+		return errors.New("invalid nonce")
 	}
 
 	// get projectID from session along with state nonce from the database
-	stateData, err := s.store.GetProjectIDPortBySessionState(ctx, in.State)
+	stateData, err := s.store.GetProjectIDPortBySessionState(ctx, state)
 	if err != nil {
-		return nil, status.Errorf(codes.Unknown, "error getting project ID by session state: %s", err)
+		return fmt.Errorf("error getting project ID by session state: %w", err)
 	}
 
-	// get provider
-	provider, err := getProviderFromRequestOrDefault(ctx, s.store, in, stateData.ProjectID)
+	err = s.generateOAuthToken(ctx, provider, code, stateData)
 	if err != nil {
-		return nil, providerError(err)
+		return err
 	}
 
+	// Telemetry logging
+	logger.BusinessRecord(ctx).Provider = provider
+	logger.BusinessRecord(ctx).Project = stateData.ProjectID
+
+	if stateData.RedirectUrl.Valid {
+		parsedURL, err := url.Parse(stateData.RedirectUrl.String)
+		if err != nil {
+			return fmt.Errorf("error parsing redirect URL: %w", err)
+		}
+		http.Redirect(w, r, parsedURL.String(), http.StatusTemporaryRedirect)
+		if err != nil {
+			return fmt.Errorf("error redirecting to URL: %w", err)
+		}
+	} else {
+		w.Header().Set("Content-Type", "text/html; charset=utf-8")
+		_, err = w.Write(auth.OAuthSuccessHtml)
+		if err != nil {
+			return fmt.Errorf("error writing OAuth success page: %w", err)
+		}
+	}
+
+	return nil
+}
+
+func (s *Server) generateOAuthToken(ctx context.Context, provider string, code string,
```

Comment:
I am sad that the existing code is entwined with the `Server` and doesn't take the cryptoEngine and a slice of the Store interface.  It would be pretty easy to test on its own if we did that.

We could also replace `DeleteAccessToken` / `CreateAccessToken` with `UpsertAccessToken`, which would do an [`INSERT ... ON CONFLICT DO UPDATE`](https://www.postgresql.org/docs/current/sql-insert.html#SQL-ON-CONFLICT), which would be more atomic than what we do today (two HTTP requests could race each other).

On the other hand, you may not be looking to do a big refactor, so I'm not going to block on it.

Reply:
I was struggling with testing this because `oauthConfig.Exchange` makes HTTP calls.
I was thinking eventually we can have a factory on the `Server` struct that generates the oauth config, similar to what we have for `auth.JwtValidator`. That would make this entire handler testable. 
I just wasn't ready to take on that refactor now. 
Do you have any other ideas for making this testable considering the calls in `oauthConfig.Exchange`?

Reply:
I like the idea, where would `Register(string, *auth.OAuthConfig)` be called? I imagine it could be on startup, since we have (almost) all the data we need to construct it. The only thing that's dynamic is the `cli` boolean which changes the redirect URL based on whether the CLI or web app initiated the enrolment, and I would like to get rid of that differentiation anyway.

Reply:
I'll create a followup issue to refactor this and make it testable